<?xml version="1.0" ?>
<robot name="baxter" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <!-- kinect with all frames as defined in turtlebot -->
   <joint name="camera_rgb_joint" type="fixed">
      <origin xyz="0 -0.045 0" rpy="0 0 0"/>
      <parent link="camera_link"/>
      <child link="camera_rgb_frame" />
  </joint>
  <link name="camera_rgb_frame"/>
  <joint name="camera_rgb_optical_joint" type="fixed">
      <origin xyz="0 0 0" rpy="-1.5707963267948966 0 -1.5707963267948966" />
      <parent link="camera_rgb_frame" />
      <child link="kinect2_link" />
  </joint>
  <link name="kinect2_link"/>
  <joint name="kinect_joint" type="fixed">
      <origin xyz="2.4 0 0.6" rpy="0 0.7 -3.14"/>
      <parent link="base"/>
      <child link="camera_link"/>
  </joint>  
  <link name="camera_link">
      <visual>
       <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
         <mesh filename="package://baxter_description/meshes/sensors/kinect.dae"/>
        </geometry>
      </visual>
  	  <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
  	    <geometry>
  	      <box size="0.07271 0.27794 0.073"/>
  	    </geometry>
  	  </collision>
      <inertial>
        <mass value="0.564" />
        <origin xyz="0 0 0" />
        <inertia ixx="0.003881243" ixy="0.0" ixz="0.0"
                 iyy="0.000498940" iyz="0.0"
                 izz="0.003879257" />
      </inertial>
  </link>
  <!-- The fixed joints & links below are usually published by static_transformers launched by the OpenNi launch 
         files. However, for Gazebo simulation we need them, so we add them here.
         (Hence, don't publish them additionally!) -->
  <joint name="camera_depth_joint" type="fixed">
	  <origin xyz="0 0.025 0" rpy="0 0 0" />
	  <parent link="camera_link" />
	  <child link="camera_depth_frame" />
  </joint>
  <link name="camera_depth_frame"/>
	<joint name="camera_depth_optical_joint" type="fixed">
	  <origin xyz="0 0 0" rpy="-1.5707963267948966 0 -1.5707963267948966" />
	  <parent link="camera_depth_frame" />
	  <child link="camera_depth_optical_frame" />
	</joint>
	<link name="camera_depth_optical_frame"/>

<!-- ros_control plugin -->
  <!--gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotNamespace>/baxter</robotNamespace>
      <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
    </plugin>
  </gazebo-->

<!-- Microsoft Kinect / ASUS Xtion PRO Live for simulation -->
    <gazebo reference="camera_link">  
      <sensor type="depth" name="camera">
        <always_on>true</always_on>
        <update_rate>20.0</update_rate>
        <camera>
          <horizontal_fov>1.05</horizontal_fov>
          <image>
            <format>B8G8R8</format>
            <width>640</width>
            <height>480</height>
          </image>
          <clip>
            <near>0.05</near>
            <far>8.0</far>
          </clip>
        </camera>
        <plugin name="kinect_camera_controller" filename="libgazebo_ros_openni_kinect.so">
          <baseline>0.2</baseline>
          <alwaysOn>true</alwaysOn>
          <updateRate>10</updateRate>
          <cameraName>kinect2</cameraName>
          <imageTopicName>rgb/image_rect_color</imageTopicName>
          <depthImageTopicName>depth_registered/sw_registered/image_rect</depthImageTopicName>
          <pointCloudTopicName>hd/points</pointCloudTopicName>
          <cameraInfoTopicName>rgb/camera_info</cameraInfoTopicName>
          <depthImageCameraInfoTopicName>depth_registered/sw_registered/camera_info</depthImageCameraInfoTopicName>
          <frameName>kinect2_link</frameName>
          <pointCloudCutoff>0.5</pointCloudCutoff>
          <distortionK1>0.00000001</distortionK1>
          <distortionK2>0.00000001</distortionK2>
          <distortionK3>0.00000001</distortionK3>
          <distortionT1>0.00000001</distortionT1>
          <distortionT2>0.00000001</distortionT2>
          <CxPrime>0</CxPrime>
          <Cx>0</Cx>
          <Cy>0</Cy>
          <focalLength>0</focalLength>
          <hackBaseline>0</hackBaseline>
        </plugin>
      </sensor>
    </gazebo>


  <xacro:arg name="gazebo" default="false"/>
  <xacro:arg name="pedestal" default="true"/>
  <!-- Baxter Base URDF -->
  <xacro:include filename="$(find baxter_description)/urdf/baxter_base/baxter_base.urdf.xacro">
    <xacro:arg name="gazebo" value="${gazebo}"/>
  </xacro:include>

  <!-- Baxter Pedestal -->
  <xacro:if value="$(arg pedestal)">
    <xacro:include filename="$(find baxter_description)/urdf/pedestal/pedestal.xacro">
      <xacro:arg name="gazebo" value="${gazebo}"/>
    </xacro:include>
  </xacro:if>

  <!-- Left End Effector -->
  <xacro:include filename="$(find baxter_description)/urdf/left_end_effector.urdf.xacro" />

  <!-- Right End Effector -->
  <xacro:include filename="$(find baxter_description)/urdf/right_end_effector.urdf.xacro" />

</robot>
